Cover: Set custom color when applying initial background image

ItsJonQ opened the pull request:
## Description

<img width="632" alt="Screen Shot 2020-05-22 at 5 27 47 PM" src="https://user-images.githubusercontent.com/2322354/82711217-bb30a480-9c52-11ea-8a54-269ed60167f4.png">

This update adds an enhancement to the Cover block for background images. When an image is first set, it extracts and the primary color of the image and applies it as a custom background color.

This is achieved by using a library [color-thief](https://github.com/lokesh/color-thief) ([2.6 kb gzipped](https://bundlephobia.com/result?p=colorthief@2.3.0)), which leverages `<Canvas />` for image parsing and color picking.

Here is a video demo of it working:
https://d.pr/v/f2flil

Here's the difference:

<img width="651" alt="Screen Shot 2020-05-22 at 5 44 42 PM" src="https://user-images.githubusercontent.com/2322354/82711645-f4b5df80-9c53-11ea-80e1-558c7f8ed486.png">


## How has this been tested?

Tested locally in Gutenberg

* Run `npm install`
* Run `npm run dev`
* Add a Cover block
* Add an image from the Media Library for the background (perhaps something colourful)
* Check to see if the background tint matches the image
* Check the Cover background color - make sure it's custom

## Types of changes

* This update adds a new `useColorExtract` hook from `@wordpress/components`, which handles the responsibility for image rendering/color extraction

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Resolves comment in:
https://github.com/WordPress/gutenberg/issues/16479#issuecomment-546547459

Commit:
Add useColorExtract hook. Set up color extract with background image in Cover.

Commit:
Fix example for useColorExtract

ItsJonQ commented:
I'm noticing a bug where we need to update the color when **changing/updating** the background image.

Will need to work on that logic next!

Commit:
Fix background color extraction to change on replace media

ItsJonQ commented:
Aww. It looks like `color-thief` has some incompatible licenses. I believe it's because the library attempts to accommodate both web and node.js. It looks like the node.js dependencies are making the license checker unhappy

Commit:
Extract core functionality from color-thief for color extraction

Commit:
Fix quantize dependency source

strarsis commented:
@ItsJonQ: What about using Gradients?
https://github.com/danieltrost/image-gradient
https://github.com/peterekepeter/image-to-gradient

The dominant color npm library got a MIT license:
https://github.com/shime/dominant-color

ItsJonQ commented:
Heya @strarsis ,

Thanks for the suggestions! I checked out the 2 libraries. We wouldn't be able to use those as those are `node.js` based. We need a web-based solution :).

Thankfully, I was able to adjust the module to fit our requirements

mtias commented:
Reminds me of https://github.com/mtias/tonesque :)

ItsJonQ commented:
@mtias Oh man. I recognize some of those RGB colour formulas 😂 (I worked on lower colour interpreting/parsing stuff in JS a couple of years ago)

mtias commented:
I like this exploration. I think it might also make sense to add a button / action to set it to "dominant color" if you have picked another one. cc @jasmussen

ItsJonQ commented:
> I think it might also make sense to add a button / action to set it to "dominant color" if you have picked another one

@mtias OoOOooo! Interesting to see what @jasmussen has for design ideas!

jasmussen commented:
For whatever reason, I couldn't get this PR to work for me, no color is stolen by the thief:

<img width="771" alt="before" src="https://user-images.githubusercontent.com/1204802/82997909-4caf6600-a007-11ea-88dd-86257f269409.png">

That said, and judging by your video, this is really really cool and I share Matías' excitement.

Very first question: can we add text color to the cover block, and set that to dark or light depending on the luminosity of the background color eye-dropped? You'd still be able to override colors on a per-paragraph basis, or even unset the text color on the cover block. But it would help create a nice contrasty basic experience. 

Transversal, but related: this tool could theoretically help the block UI. We currently know whether a theme has a light or dark background because the theme can [tell us](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#dark-backgrounds), however even then we have cases where the background color is sub-optimal for the block UI. Like this blue group with a light-gray spacer inside:

<img width="783" alt="Screenshot 2020-05-27 at 10 47 51" src="https://user-images.githubusercontent.com/1204802/82998301-da8b5100-a007-11ea-9516-3110e7dea197.png">

Back to the PR in question, this is so lovely, it makes the end result look much better. It reminds me of the "faded" filtered images that are popular on Instagram, but more than that it's just a better overlay with the text as a background.

I'll put together a small design tool iteration for the cover block sidebar that includes a "pick dominant color" button — for the time being though, feel free to just go ahead and make that button if you can, because I agree with Matías, it'd be nice to be able to pick this after the fact too. I might want to go back to an old cover block I had and set this, or I might explore other colors and then want to go back to the auto-picked one.

jasmussen commented:
Here's a quick experiment:

<img width="1250" alt="Frame 146" src="https://user-images.githubusercontent.com/1204802/83118215-5ea80c00-a0ce-11ea-9fe4-de67baf41489.png">

As with many of these G2 designs, they rely on some designs landing in the sidebar that aren't there yet. Please consider this like a destination we'll work towards.

ItsJonQ commented:
@jasmussen Thanks for the feedback!

> ...no color is stolen by the thief:

Oh darn! Do you think you can forward me a link to that image? Probably worth test it (on my end) using the same one that's giving you issues.

> Very first question: can we add text color to the cover block, and set that to dark or light depending on the luminosity of the background color eye-dropped?

You bet! I think that would be fun! Probably a little tricky... but worth trying!

> ...it'd be nice to be able to pick this after the fact too. I might want to go back to an old cover block I had and set this, or I might explore other colors and then want to go back to the auto-picked one.

I'll give this a shot too! Sounds exciting :D

jasmussen commented:
Very curious, for a moment I thought it was a missing npm install I did, but that didn't help. I still think it's possibly something on my end, but I still can't get it to work. The pictures I keep using are these: https://cloudup.com/cAYtklfVZtk

> You bet! I think that would be fun! Probably a little tricky... but worth trying!

Not necessary for this PR at all, just an idea!

Commit:
Fix color extract on initial image setting

ItsJonQ commented:
@jasmussen I think I fixed it! Totally wasn't your fault. There's some tricky logic I needed to add 🙃 - mostly around initial load and tracking `src` updates.

Let me know if it works!

P.S. Great mountain pics! 🗻

ItsJonQ commented:
> As with many of these G2 designs, they rely on some designs landing in the sidebar that aren't there yet. Please consider this like a destination we'll work towards.

@jasmussen From a code perspective, it's quite difficult (or rather, tricky/messy) to add a "Pick Dominant" action anywhere in this area at the moment:

<img width="246" alt="Screen Shot 2020-06-02 at 10 26 19 AM" src="https://user-images.githubusercontent.com/2322354/83531879-8676e580-a4bb-11ea-81fd-7b27af86f7b1.png">

If there were no limitations... I could imagine having a 3rd button in the SegmentControl:

```
[ Solid | Gradient | Image ]
```

(Can replace "Image" with another label)

Within the "Image" tab, you'd see a palette of colour values extracted from the image.

This **could** be something we could try, if you think the idea/direction feels good.

Either way, if we need to do anything with that Color settings area... it'll be messy 😛

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Add "Pick dominant" button in Overlay InspectorControl

ItsJonQ commented:
@jasmussen Haii! I've added the feature to "Pick dominant" on demand via a Button.

The design/placement isn't the best, but it's a start! And it's functionally working

<img width="257" alt="Screen Shot 2020-06-02 at 2 03 05 PM" src="https://user-images.githubusercontent.com/2322354/83554353-9f42c380-a4da-11ea-8ca1-094259bc38e3.png">

Any design thoughts on this would be much appreciated 🙇

gziolo commented on packages/components/src/utils/hooks/extract-color.js:
```diff
@@ -0,0 +1,211 @@
+};
+
+ColorThief.prototype.getImageData = function ( imageUrl, callback ) {
+	const xhr = new XMLHttpRequest();
```
Would it be possible to avoid a network request and process the content of file when it's uploaded? In some cases like drag & drop or when uploading with the media placeholder (the one in Gutenberg) we already use data of the image to show the placeholder. /cc @ellatrix

mcsf replied:
Tapping into `getMedia` could be good.

ItsJonQ replied:
I tried to not touch this file 🙈 . I'm open to changing this implementation if it's a concern

jasmussen commented:
Happy news, I got the color thing to work, and it is delicious:

<img width="1070" alt="Screenshot 2020-06-03 at 08 53 27" src="https://user-images.githubusercontent.com/1204802/83605148-b4216480-a577-11ea-9a0e-53b249243d81.png">

Thank you Q! If I'm not wrong, this should make it possible to reach this design, even if the button is outside the containing color picker element:

<img width="1250" alt="Frame 146" src="https://user-images.githubusercontent.com/1204802/83604761-fa29f880-a576-11ea-8131-3fbf5fa0822b.png">

Or, at the very least, this:

<img width="1250" alt="Frame 147" src="https://user-images.githubusercontent.com/1204802/83604765-fd24e900-a576-11ea-9ca2-26a5bcef2a62.png">

I know we're not quite there yet. 

Can we do this?

<img width="301" alt="Screenshot 2020-06-03 at 09 01 00" src="https://user-images.githubusercontent.com/1204802/83605775-c8199600-a578-11ea-80bf-d64c54b45f8e.png">

That gets us closer to the destination.

mtias commented:
I think "Pick dominant" can be a bit obscure (dominant form where?). What if we pass it as a new color to the palette, could that work?

jasmussen commented:
Nice idea. We'd need a special treatment for the swatch (perhaps a use case for the asterisk icon? Or maybe just the unread dot?) and a good tooltip.

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Add custom dominant color to Solid color palette

This update removes the "Pick dominant" button in favour of merging the
custom extract color alongside the site's color palette.

ItsJonQ commented:
<img width="265" alt="Screen Shot 2020-06-03 at 12 55 16 PM" src="https://user-images.githubusercontent.com/2322354/83665667-bd232d80-a599-11ea-8819-69d0c11ec0ae.png">

@mtias  + @jasmussen Thanks for the ideas!

I've adjusted things so that the extracted dominant color is merged with the color palette in the inspector controls. It's currently placed in the first position. It seems to work okay!

Let me know how it feels for you folks :)

jasmussen commented:
Wonderful, this seems like a solid path! 

Should we try this?

<img width="1250" alt="Frame 148" src="https://user-images.githubusercontent.com/1204802/83722010-82260600-a63c-11ea-9eba-b7b42d9d1d09.png">

<img width="1250" alt="Frame 149" src="https://user-images.githubusercontent.com/1204802/83722013-83efc980-a63c-11ea-97b5-a19c4bc1e4fe.png">

I'm not completely convinced myself, based on the static mockup, but it's a thing that might feel natural when in the code. The blue dot is a new interface element we use in various places to indicate "there's something here", or "this is special" or "you have unsaved changes" — a way to bring attention, just like a notification dot in other words. It's used already in the FSE experience to indicate unsaved template changes, and we'd want to use the same element and treatment, of course.

mtias commented:
Nice, I think this could work. It might make sense to add some plain text right above the color palette saying the default / first color is dynamically grabbed from the image.

> Update: I'm not sure what the label for the tooltip should be. At the moment, I just have "Dominant color" (I feel like this could be improved)

If we want to be short: "Image dominant color". If there's room for more text "Dominant color extracted from the background image".

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Update Image dominant color tooltip label

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Fix test from merge issue

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Fix E2E tests. Adjust try/catch + only loading if src in extractColor method

Commit:
Wrap getImageNode() in try/catch to handle CORS error

jasmussen commented:
I see this:

![cover](https://user-images.githubusercontent.com/1204802/86348723-58bec100-bc60-11ea-8d86-401180a44dbc.gif)

It feels good. I think we can still think about some indication that the added color is magically added, like a dot or something. But I think there's such value in this that I'd be fine with merging as is and improving design tools separately.

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
Fix extract function to run only for Editor (not preview)

mcsf commented on packages/block-library/src/cover/edit.js:
```diff
@@ -542,6 +559,53 @@ function CoverEdit( {
+	if ( customExtractedColor ) {
+		customColors = [
+			{
+				name: __( 'Image dominant color' ),
+				slug: __( 'image-dominant-color' ),
```
Some context or `translators:` notes could be good for these.

mcsf commented on packages/components/src/index.js:
```diff
@@ -131,3 +131,6 @@ export {
 export { default as withSpokenMessages } from './higher-order/with-spoken-messages';
+
+// Utilities
+export { useColorExtract, useControlledState } from './utils/hooks';
```
Better address `useControlledState` in a separate PR, otherwise it'll be harder in the future to trace changes back to discussions.

mcsf reviewed (commented):
I'm happy to see those passing e2e tests following cec9121ffb40cb933e4be8f9da7c987acfa0548e :)

Commit:
Remove export of useControlledState

jorgefilipecosta commented on packages/block-library/src/cover/edit.js:
```diff
@@ -1,6 +1,7 @@
  */
+import { noop } from 'lodash';
 import classnames from 'classnames';
 import FastAverageColor from 'fast-average-color';
```
Hi @ItsJonQ, the cover block was already computing the color of an image using the FastAverageColor module. I think we should avoid using two different modules for the same task so I guess we should opt for one of the modules and update the code to use the chosen module.

ItsJonQ replied:
@jorgefilipecosta Oh! I wasn't aware of that. Thank you for pointing that out. I'll make an update

Commit:
Merge branch 'master' into try/cover-background-image-auto-color

Commit:
useExtractColor: Replace ColorThief with FastAverageColor module (already being used in Block/Cover)

Commit:
Adjust fast-average-color package to match the one used in block-library

annezazu commented:
Closing this out due to inactivity :)

mtias commented:
@annezazu I'd still love to try something like this. Can we transfer it to an issue?

annezazu commented:
Done :)
